s3/pkg/service/object.go

```diff
@@ -155,6 +158,15 @@ func (s *s3Service) PutObject(ctx context.Context, in pb.S3_PutObjectStream) err
 		limitedDataReader = data
 	}
 
+	// encrypt if needed
+	if bucket.ServerSideEncryption.SseType == "SSE" {
+		byteArr, _ := ioutil.ReadAll(limitedDataReader)
+		key, _ := utils.GetRandom32BitKey()
+		_, encBuf := utils.EncryptWithAES256RandomKey(byteArr, key)
```

Comment:
Hi @rhsakarpos Does this mean it will read all of the input and put it to a buffer, and then encrypt it and save to a new buffer? If it is like that, then that would need too much memory.

Reply:
yes @sfzeng right, it will take double the initial memory. opening issue to track this, will implement your idea of a wrapped reader, which should do this ?
_The wrapped reader looks good, i think this will save memory ? as in, it will read and encrypt on the fly ? as against read-all-make-a-copy-then-encrypt-all ?_
Same goes for decrypt too, as mentioned in #788